server/src/main/java/org/elasticsearch/ingest/IngestDocMetadata.java

```diff
@@ -42,6 +42,20 @@ class IngestDocMetadata extends Metadata {
         new FieldProperty<>(Map.class).withWritable().withNullable()
     );
 
+    private static final char UNDERSCORE = '_';
+    static {
+        // there's an optimization here in the overridden isAvailable below, but it only works if the first character of each of these
+        // keys starts with an underscore, since we know all the keys up front, though, we can just make sure that's always true
+        validateLeadingUnderscores(PROPERTIES);
+    }
+
+    private static Map<String, FieldProperty<?>> validateLeadingUnderscores(final Map<String, FieldProperty<?>> properties) {
+        for (String key : properties.keySet()) {
+            assert key.charAt(0) == UNDERSCORE;
```

Comment:
@masseyke `assert` made sense to me before I pulled this out into a method, now I'm thinking it should be an `IllegalArgumentException`, perhaps. What do you think?

Reply:
I switched it in a683c54dd4f8ad18c2f8a3555d968585472e0846.

Reply:
It looks like the constructor that lets you pass in arbitrary properties is only used by unit tests. It seems a little odd that we have this dangerous constructor and validation around it just for the sake of unit tests.

Reply:
c090cbe62fb4b5c3e8f649e4f57c1e5a9bbcdf28 drops the bad constructor and fusses with a handful of tests that were using the more flexible version of things. I always hated `TestIngestCtxMetadata` so thanks for pushing on this, because now it's gone.